Panos network operations objects

adambaumeister opened the pull request:
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [X] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
The final set of commands to be used by the network operations pack! This set focuses on commit tasks but it also has a helper for pulling objects from the firewalls using pan-os-python which we can expand in the future to be more feature-rich.

@GuyAfik the only thing this PR is missing is updates to the test-playbooks. Long story but my lab is out of action over the weekend, if you could add these five commands to the playbooks using `demisto-sdk generate-test-playbook -i Panorama.yml -e <file containing the five new commands>` it would save me a lot of time. If not it's ok, I can do on monday instead.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [X] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [X] No

## Must have
- [X] Tests
- [X] Documentation

GuyAfik reviewed (commented):
Heya @adambaumeister, ill for sure add those commands to the test-playbooks, thank you for your massive efforts!

See my comments, for any question as usual lets slack it up!

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -9,7 +10,10 @@
 from panos.panorama import Panorama, DeviceGroup, Template, PanoramaCommitAll
 from panos.policies import Rulebase, PreRulebase, PostRulebase, SecurityRule
-from panos.objects import LogForwardingProfile, LogForwardingProfileMatchList
+from panos.objects import (
+    LogForwardingProfile, LogForwardingProfileMatchList, AddressObject, AddressGroup, ServiceObject, ServiceGroup,
+    ApplicationObject, ApplicationGroup, SecurityProfileGroup
+)
```
Fixed this - I was using globals() as a shortcut in some code but I replaced it with a simple dictionary to make it clearer.

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -9979,15 +10065,17 @@ def get_bgp_peers(
 
     @staticmethod
-    def get_device_state(topology: Topology, device_filter_str: str):
+    def get_device_state(topology: Topology, target: str):
```
We hadn't implemented the actual command yet but we had implemented this function earlier - this PR actually adds the command itself so to keep it consistent with our other operational commands I added the `target` argument. Can't be backwards compatibility issue as the command not yet implmented.

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -8896,6 +8901,81 @@ def fix_security_rule_security_profile_group(topology: Topology,
+                container.add(firewall_rulebase)
+                unfiltered_objects = object_class.refreshall(firewall_rulebase)
+                unfiltered_objects += object_class.refreshall(pre_rulebase)
+                unfiltered_objects += object_class.refreshall(post_rulebase)
```
Agreed, I split this into a new function that supports passing the object type in.

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -10584,6 +10672,119 @@ def fix_security_rule_security_profile_group(
+    return PanoramaCommand.push_all(
+        topology,
+        device_filter_string,
+        device_group_filter=device_group_filter,
+        template_stack_filter=template_stack_filter
```
Makes snese- fixed

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.yml:
```diff
@@ -8117,6 +8117,226 @@ script:
+      target to get the device state - if the IP address, as reported in 'show system info' is unreachable, this command will
+      fail.
+    name: pan-os-platform-get-device-state
+    outputs:
+    - contextPath: InfoFile.Name
+      description: Filename
+      type: String
+    - contextPath: InfoFile.EntryID
+      description: Entry ID
+      type: String
+    - contextPath: InfoFile.Size
+      description: Size of file
+      type: String
+    - contextPath: InfoFile.Type
+      description: Type of file
+      type: String
+    - contextPath: InfoFile.Info
+      description: Basic information of file
+      type: String
```
👍

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -8896,6 +8901,81 @@ def fix_security_rule_security_profile_group(topology: Topology,
+        :param object_name: The name of the object to find; can be regex if use_regex is set
+        :param use_regex: Whether we should use regex matching for the object_name
+        """
+        object_class = globals().get(object_type, None)
```
I changed this - was a convienience but it could introduce a problem down the track.

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -9543,49 +9623,55 @@ def push_all(
+                        commit_all_status="Initiated",
+                        device_status="",
+                        device=""
+                    ))
```
Agreed - this simplified the code a lot. Done!

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -8812,24 +8817,28 @@ def fix_security_zone_no_log_setting(
-        Gets all the security rule objects from the given VSYS or device group and return them.
-        :param container: The object to search for the rules in, as passed to pan-os-python
-        """
+    def get_all_rules_in_container(container: Union[Panorama, Firewall, DeviceGroup, Template, Vsys],
+                                   object_class: Any):
```
would you be able to mention what kind of objects could ```object_class``` be? 
also add a short doc-string

adambaumeister replied:
Ah yes, I missed this.

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -8812,24 +8817,28 @@ def fix_security_zone_no_log_setting(
-        """
+    def get_all_rules_in_container(container: Union[Panorama, Firewall, DeviceGroup, Template, Vsys],
+                                   object_class: Any):
         firewall_rulebase = Rulebase()
         pre_rulebase = PreRulebase()
         post_rulebase = PostRulebase()
         container.add(pre_rulebase)
         container.add(post_rulebase)
         container.add(firewall_rulebase)
```
```suggestion
        if object_class is not (SecuirtyRule, NatRule, ....):
            raise ValueError(f'{object_class} must be of type: ....)

        firewall_rulebase = Rulebase()
        pre_rulebase = PreRulebase()
        post_rulebase = PostRulebase()
        container.add(pre_rulebase)
        container.add(post_rulebase)
        container.add(firewall_rulebase)
```

this is not a must, its just to be on the safe side here.

adambaumeister replied:
Good idea

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -8896,6 +8905,84 @@ def fix_security_rule_security_profile_group(topology: Topology,
+            "SecurityRule": SecurityRule,
+            "NatRule": NatRule,
+        }
+        object_class = supported_object_types.get(object_type, None)
```
```suggestion
        object_class = supported_object_types.get(object_type)
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -8896,6 +8905,84 @@ def fix_security_rule_security_profile_group(topology: Topology,
+        :param object_name: The name of the object to find; can be regex if use_regex is set
+        :param use_regex: Whether we should use regex matching for the object_name
+        """
+        supported_object_types = {
+            "AddressObject": AddressObject,
+            "AddressGroup": AddressGroup,
+            "ServiceObject": ServiceObject,
+            "ServiceGroup": ServiceGroup,
+            "ApplicationObject": ApplicationObject,
+            "ApplicationGroup": ApplicationGroup,
+            "SecurityProfileGroup": SecurityProfileGroup,
+            "SecurityRule": SecurityRule,
+            "NatRule": NatRule,
+        }
```
consider making this as a class member

adambaumeister replied:
Good idea, done

GuyAfik reviewed (commented):
@adambaumeister nicely done, just a few additional minor comments
Lets setup a demo

GuyAfik commented:
@adambaumeister FYI pushed the new commands to our test-playbooks :)

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -8896,6 +8914,86 @@ def fix_security_rule_security_profile_group(topology: Topology,
+class ObjectGetter:
+    """Retrieves objects from the PAN-OS configuration"""
+
+    supported_object_types = {
```
```suggestion
    SUPPORTED_OBJECT_TYPES = {
```

GuyAfik reviewed (commented):
@adambaumeister small comment on typing, then just demo and we are done

adambaumeister commented:
Ok so I said it was gonna take me a couple of days but I just knocked it out before I could forget what we spoke about.

Added 
 * pan-os-push-to-template
 * pan-os-push-to-template-stack

I also deleted the other pan-os-python based commands for managing the commit as we should now have captured all the requirements for panorama configurations. They may not implement every bit of functionality the API provides for these but we can expand them further down the line.

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -730,6 +801,66 @@ def panorama_push_to_device_group_command(args: dict):
+            'Status': 'Pending'
+        }
+        return_results({
+            'Type': entryTypes['note'],
+            'ContentsFormat': formats['json'],
+            'Contents': result,
+            'ReadableContentsFormat': formats['markdown'],
+            'HumanReadable': tableToMarkdown('Push to Template:', push_output, ['JobID', 'Status'],
+                                             removeNull=True),
+            'EntryContext': {
+                "Panorama.Push(val.JobID == obj.JobID)": push_output
+            }
```
is there a reason you don't use ```CommandResults```? or is it mainly because that is the way its implemented in most of the integration? I'd be happy to see you using ```CommandResults``` instead

adambaumeister replied:
I think it's a little simpler to keep each of these functions looking the same as they are all really similar

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -730,6 +801,66 @@ def panorama_push_to_device_group_command(args: dict):
+            'Status': 'Pending'
+        }
+        return_results({
+            'Type': entryTypes['note'],
+            'ContentsFormat': formats['json'],
+            'Contents': result,
+            'ReadableContentsFormat': formats['markdown'],
+            'HumanReadable': tableToMarkdown('Push to Template:', push_output, ['JobID', 'Status'],
+                                             removeNull=True),
+            'EntryContext': {
+                "Panorama.Push(val.JobID == obj.JobID)": push_output
+            }
```
same as above

adambaumeister replied:
same as above comment - easier to keep them the same

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -698,6 +702,73 @@ def panorama_push_to_device_group(args: dict):
+    command: str = ''
+    command += f'<name>{TEMPLATE}</name>'
+
+    serial_number = args.get('serial_number')
+    if serial_number:
+        command = f'<name>{TEMPLATE}</name><device><member>{serial_number}</member></device>'
```
```suggestion
    if serial_number := args.get('serial_number')
        command = f'<name>{TEMPLATE}</name><device><member>{serial_number}</member></device>'
```

adambaumeister replied:
Fixed

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -698,6 +702,73 @@ def panorama_push_to_device_group(args: dict):
+    command: str = ''
+    command += f'<name>{template_stack}</name>'
+
+    serial_number = args.get('serial_number')
+    if serial_number:
+        command = f'<name>{template_stack}</name><device><member>{serial_number}</member></device>'
```
```suggestion
    if serial_number := args.get('serial_number')
        command = f'<name>{template_stack}</name><device><member>{serial_number}</member></device>'
```

adambaumeister replied:
Fixed

GuyAfik reviewed (commented):
Nicely done @adambaumeister, can you add unit-tests for the following commands?
* ```pan-os-push-to-template```
* ```pan-os-push-to-template-stack```

Ill update the test-playbooks accordingly

Other than that looks really good!

adambaumeister commented:
Added unit tests and corrected those minor things!

GuyAfik commented:
@jrosenpanw can you set someone up to do a doc-review here?

GuyAfik reviewed (approved):
@adambaumeister excellent work, fix the small build issues if you are able, only flake8 issues.

ShirleyDenkberg commented:
@GuyAfik Doc review completed.
